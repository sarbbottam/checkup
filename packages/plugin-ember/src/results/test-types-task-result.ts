import { BaseTaskResult, TaskItemData, TaskResult, ui, CardData } from '@checkup/core';

export default class TestTypesTaskResult extends BaseTaskResult implements TaskResult {
  testTypes!: TaskItemData[];

  findByType(typeName: string): TaskItemData | undefined {
    return this.testTypes.find(type => type.type === typeName);
  }

  stdout() {
    ui.styledHeader(this.meta.friendlyTaskName);
    ui.blankLine();
    ui.table(this._deriveTestPercentage(this.testTypes), { type: {}, total: {}, percentage: {} });
    ui.blankLine();
  }

  json() {
    return { meta: this.meta, result: { types: this.testTypes } };
  }

  pdf() {
    return new CardData(this.meta, 20, 50, 25, 'tests left to migrate');
  }

  _deriveTestPercentage(testTypes: TaskItemData[]) {
    const totalTests: number = testTypes.reduce(
      (accum, currentValue) => accum + currentValue.total,
      0
    );
    return testTypes.map(testType => {
      testType.percentage = this._formatAsPercetange(testType.total / totalTests);
      return testType;
    });
  }

  _formatAsPercetange(value: number) {
    return `${(value * 100).toFixed()}%`;
  }
}
