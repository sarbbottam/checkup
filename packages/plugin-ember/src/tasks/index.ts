export { default as EmberProjectTask } from './ember-project-task';
export { default as DependenciesTask } from './dependencies-task';
export { default as TypesTask } from './types-task';
export { default as TestTypesTask } from './test-types-task';
