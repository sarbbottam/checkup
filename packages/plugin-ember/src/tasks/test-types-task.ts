import { Task, Category, FileSearcherTask, Priority, TaskResult } from '@checkup/core';

import { TestTypesTaskResult } from '../results';

// NOTE: this task gets number of test files, not tests. This will be iterated on and changed in the future TODO: @ckessler
const SEARCH_PATTERNS = {
  unit: ['**/unit/**/*-test.js'],
  integration: ['**/integration/**/*-test.js'],
  acceptance: ['**/acceptance/**/*-test.js'],
};

export default class TestTypesTask extends FileSearcherTask implements Task {
  meta = {
    taskName: 'test-types',
    friendlyTaskName: 'Test Types',
    taskClassification: {
      category: Category.Core,
      priority: Priority.Medium,
    },
  };

  constructor(cliArguments: any) {
    super(cliArguments, SEARCH_PATTERNS);
  }

  async run(): Promise<TaskResult> {
    let result = new TestTypesTaskResult(this.meta);
    result.testTypes = await this.searcher.search();

    return result;
  }
}
