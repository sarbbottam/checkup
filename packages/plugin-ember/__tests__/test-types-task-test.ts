import { EmberProject, stdout } from '@checkup/test-helpers';

import { TestTypesTask } from '../src/tasks';
import { TestTypesTaskResult } from '../src/results';

const TESTS = {
  acceptance: {
    tests: {
      'my-component-test.js': '',
    },
    someFolder: {
      'my-component-test.js': '',
    },
  },
  integration: {
    tests: {
      'my-component-test.js': '',
    },
  },
  unit: {
    tests: {
      'my-component-test.js': '',
    },
  },
};

describe('test-types-task', () => {
  let project: EmberProject;

  beforeEach(function() {
    project = new EmberProject('checkup-app', '0.0.0');
  });

  afterEach(function() {
    project.dispose();
  });

  it('returns all the test types found in the app and outputs to the console', async () => {
    project.files = Object.assign(project.files, {
      'index.js': 'index js file',
      addon: TESTS,
    });

    project.writeSync();

    const result = await new TestTypesTask({ path: project.baseDir }).run();
    const testTypesTaskResult = <TestTypesTaskResult>result;

    testTypesTaskResult.stdout();

    expect(stdout()).toMatchSnapshot();
  });

  it('returns all the test types (including nested) found in the app and outputs to the console', async () => {
    project.files = Object.assign(project.files, {
      'index.js': 'index js file',
      addon: TESTS,
    });

    project.addInRepoAddon('ember-super-button', 'latest');

    // @ts-ignore
    project.files.lib['ember-super-button'].addon = TESTS;
    // @ts-ignore

    project.writeSync();

    const result = await new TestTypesTask({ path: project.baseDir }).run();
    const testTypesTaskResult = <TestTypesTaskResult>result;

    testTypesTaskResult.stdout();

    expect(stdout()).toMatchSnapshot();
  });

  it('returns all the test types found in the app and outputs to JSON', async () => {
    project.files = Object.assign(project.files, {
      'index.js': 'index js file',
      addon: TESTS,
    });

    project.writeSync();

    const result = await new TestTypesTask({ path: project.baseDir }).run();
    const testTypesTaskResult = <TestTypesTaskResult>result;

    expect(testTypesTaskResult.json()).toMatchSnapshot();
  });

  it('returns all the test types (including nested) found in the app and outputs to JSON', async () => {
    project.files = Object.assign(project.files, {
      'index.js': 'index js file',
      addon: TESTS,
    });

    project.addInRepoAddon('ember-super-button', 'latest');

    // @ts-ignore
    project.files.lib['ember-super-button'].addon = TESTS;
    // @ts-ignore

    project.writeSync();

    const result = await new TestTypesTask({ path: project.baseDir }).run();
    const testTypesTaskResult = <TestTypesTaskResult>result;

    expect(testTypesTaskResult.json()).toMatchSnapshot();
  });
});
