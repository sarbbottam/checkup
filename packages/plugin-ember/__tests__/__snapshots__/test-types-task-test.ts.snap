// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-types-task returns all the test types (including nested) found in the app and outputs to JSON 1`] = `
Object {
  "meta": Object {
    "friendlyTaskName": "Test Types",
    "taskClassification": Object {
      "category": "core",
      "priority": "medium",
    },
    "taskName": "test-types",
  },
  "result": Object {
    "types": Array [
      Object {
        "data": Array [
          "addon/unit/tests/my-component-test.js",
          "lib/ember-super-button/addon/unit/tests/my-component-test.js",
        ],
        "total": 2,
        "type": "unit",
      },
      Object {
        "data": Array [
          "addon/integration/tests/my-component-test.js",
          "lib/ember-super-button/addon/integration/tests/my-component-test.js",
        ],
        "total": 2,
        "type": "integration",
      },
      Object {
        "data": Array [
          "addon/acceptance/someFolder/my-component-test.js",
          "addon/acceptance/tests/my-component-test.js",
          "lib/ember-super-button/addon/acceptance/someFolder/my-component-test.js",
          "lib/ember-super-button/addon/acceptance/tests/my-component-test.js",
        ],
        "total": 4,
        "type": "acceptance",
      },
    ],
  },
}
`;

exports[`test-types-task returns all the test types (including nested) found in the app and outputs to the console 1`] = `
"=== Test Types

Type        Total Percentage 
unit        2     25%        
integration 2     25%        
acceptance  4     50%        

"
`;

exports[`test-types-task returns all the test types found in the app and outputs to JSON 1`] = `
Object {
  "meta": Object {
    "friendlyTaskName": "Test Types",
    "taskClassification": Object {
      "category": "core",
      "priority": "medium",
    },
    "taskName": "test-types",
  },
  "result": Object {
    "types": Array [
      Object {
        "data": Array [
          "addon/unit/tests/my-component-test.js",
        ],
        "total": 1,
        "type": "unit",
      },
      Object {
        "data": Array [
          "addon/integration/tests/my-component-test.js",
        ],
        "total": 1,
        "type": "integration",
      },
      Object {
        "data": Array [
          "addon/acceptance/someFolder/my-component-test.js",
          "addon/acceptance/tests/my-component-test.js",
        ],
        "total": 2,
        "type": "acceptance",
      },
    ],
  },
}
`;

exports[`test-types-task returns all the test types found in the app and outputs to the console 1`] = `
"=== Test Types

Type        Total Percentage 
unit        1     25%        
integration 1     25%        
acceptance  2     50%        

"
`;
